subsys/net/lib/app/client.c

```diff
@@ -639,6 +639,44 @@ int net_app_connect(struct net_app_ctx *ctx, s32_t timeout)
 		ctx->is_enabled = true;
 
 		_net_app_print_info(ctx);
+	} else {
+		/* We cannot bind to local unspecified address when sending.
+		 * Select proper address depending on remote one in this case.
+		 */
+#if defined(CONFIG_NET_IPV6)
+		if (net_context_get_family(net_ctx) == AF_INET6 &&
+		    net_is_ipv6_addr_unspecified(
+			    &net_sin6(&ctx->ipv6.local)->sin6_addr)) {
+			const struct in6_addr *addr;
+
+			addr = net_if_ipv6_select_src_addr(
+				NULL,
+				&net_sin6(&ctx->ipv6.remote)->sin6_addr);
+			if (addr && addr != net_ipv6_unspecified_address()) {
+				net_ipaddr_copy(
+					&net_sin6(&ctx->ipv6.local)->sin6_addr,
+					addr);
+			}
+		}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+		if (net_context_get_family(net_ctx) == AF_INET &&
+		    net_is_ipv4_addr_unspecified(
+			    &net_sin(&ctx->ipv4.local)->sin_addr)) {
+			struct net_if *iface;
+
+			/* Just take the first IPv4 address of an interface */
+			iface = net_context_get_iface(net_ctx);
+			if (iface) {
+				net_ipaddr_copy(
+				      &net_sin(&ctx->ipv4.local)->sin_addr,
+				      &iface->ipv4.unicast[0].address.in_addr);
+			} else {
+				NET_WARN("Source address is unspecified!");
+			}
+		}
+#endif
```

Comment:
Could you move this code into a dedicated function? With a nice name that self-describes what it does (easy to say though... ^^)

Current one is already quite lengthy, and with all the if imbrication, it is now giving weird indentation as in lines 653, 654.

Reply:
Sure, np